base/abstractarray.jl

```diff
@@ -2506,6 +2506,255 @@ end
+
+    iit = iterate(newaxis)
+    while xit !== nothing
+        x, state = xit
+        i, istate = iit
+        _stack_size_check(x, ax1)
+        @inbounds if dims==1
+            inds1 = ntuple(d -> d==1 ? i : Colon(), N1)
+            if x isa AbstractArray
+                B[inds1...] = x
+            else
+                copyto!(view(B, inds1...), x)
+            end
+        elseif dims==2
+            inds2 = ntuple(d -> d==2 ? i : Colon(), N1)
+            if x isa AbstractArray
+                B[inds2...] = x
+            else
+                copyto!(view(B, inds2...), x)
+            end
+        else
+            inds = ntuple(d -> d==dims ? i : Colon(), N1)
+                if x isa AbstractArray
+                B[inds...] = x
+            else
+                # This is where the type-instability of inds hurts, but it is pretty exotic:
+                copyto!(view(B, inds...), x)
+            end
+        end
+        xit = iterate(A, state)
+        iit = iterate(newaxis, istate)
+    end
+    B
+end
+
+@inline function _stack_size_check(x, ax1::Tuple)
+    if _axes(x) != ax1
+        uax1 = UnitRange.(ax1)
+        uaxN = UnitRange.(axes(x))
+        throw(DimensionMismatch(
+            "stack expects uniform slices, got axes(x) = $uaxN while first had $uax1"))
+    end
+end
+
+# For `similar`, the goal is to stack an Array of CuArrays to a CuArray:
+_prototype(x::AbstractArray, A::AbstractArray) = x
+_prototype(x::AbstractArray, A) = x
+_prototype(x, A::AbstractArray) = A
+_prototype(x, A) = 1:0
+
+# With tuple elements, we can make the empty array the right size:
+function _empty_stack(::Colon, ::Type{T}, ::Type{S}, A) where {T, S<:Tuple}
+    similar(_prototype(nothing, A), T, OneTo(length(fieldtypes(S))), axes(A)...)
+end
+function _empty_stack(dims::Integer, ::Type{T}, ::Type{S}, A) where {T, S<:Tuple}
+    ax1 = OneTo(length(fieldtypes(S)))
+    dims in 1:2 || throw(ArgumentError("cannot stack tuples along dims = $dims"))
+    similar(_prototype(nothing, A), T, ntuple(d -> d==dims ? OneTo(0) : ax1, 2))
+end
+# but with arrays of arrays, we must settle for the right ndims:
+_empty_stack(dims, ::Type{T}, ::Type{S}, A) where {T,S} = _empty_stack(dims, T, IteratorSize(S), A)
+_empty_stack(dims, ::Type{T}, ::HasLength, A) where {T} = _empty_stack(dims, T, HasShape{1}(), A)
+_empty_stack(dims, ::Type{T}, ::IteratorSize, A) where {T} = _empty_stack(dims, T, HasShape{0}(), A)
+
+function _empty_stack(::Colon, ::Type{T}, ::HasShape{N}, A) where {T,N}
+    similar(_prototype(nothing, A), T, ntuple(_->OneTo(1), N)..., _axes(A)...)
+end
+function _empty_stack(dims::Integer, ::Type{T}, ::HasShape{N}, A) where {T,N}
+    # Not sure we should check dims here, e.g. stack(Vector[]; dims=2) is an error
+    dims in 1:N+1 || throw(ArgumentError("cannot stack slices ndims(x) = $N along dims = $dims"))
+    ax = ntuple(d -> d==dims ? _vec_axis(A) : OneTo(1), N+1)
+    similar(_prototype(nothing, A), T, ax...)
+end
```

Comment:
Some empty behaviours, most to least specific:
```
julia> stack(Tuple{Int,Int,Int}[])
3×0 Matrix{Int64}

julia> stack([(1,2,3)])
3×1 Matrix{Int64}:
 1
 2
 3

julia> stack(Matrix{Int}[])
1×1×0 Array{Int64, 3}

julia> stack([[1 2; 3 4]])
2×2×1 Array{Int64, 3}:
[:, :, 1] =
 1  2
 3  4

julia> stack(Matrix[])  # eltype is abstract, perhaps not ideal?
Any[]

julia> stack([])
Any[]

julia> stack([1,2,3])  # because numbers are iterable
3-element Vector{Int64}:
 1
 2
 3
```
My thinking re the Tuple handling is partly that SVectors could share these paths. It's possible I haven't thought enough about SMatrices.